src/components/db-explorer.tsx

```diff
@@ -0,0 +1,249 @@
+import { sql } from "@codemirror/lang-sql";
+import * as duckdb from "@duckdb/duckdb-wasm";
+import eh_worker from "@duckdb/duckdb-wasm/dist/duckdb-browser-eh.worker.js?url";
+import duckdb_wasm_next from "@duckdb/duckdb-wasm/dist/duckdb-eh.wasm?url";
+import duckdb_wasm from "@duckdb/duckdb-wasm/dist/duckdb-mvp.wasm?url";
+import { Grid } from "@githubocto/flat-ui";
+import CodeMirror from "@uiw/react-codemirror";
+import React, { useEffect, useMemo, useRef, useState } from "react";
+import { ErrorBoundary, FallbackProps } from "react-error-boundary";
+import { useQuery } from "react-query";
+import { useDebounce } from "use-debounce";
+import Bug from "../bug.svg";
+import { useDataFile } from "../hooks";
+import { ErrorState } from "./error-state";
+import { LoadingState } from "./loading-state";
+import { Spinner } from "./spinner";
+
+interface Props {
+  sha: string;
+  filename: string;
+  owner: string;
+  name: string;
+}
+
+interface DBExplorerInnerProps {
+  content: string;
+  filename: string;
+  extension: string;
+  sha: string;
+}
+
+const VALID_EXTENSIONS = ["csv", "json"];
+
+function ErrorFallback(props: FallbackProps) {
+  const { error, resetErrorBoundary } = props;
+  return (
+    <ErrorState img={Bug} alt={error.message}>
+      <p>{error?.message}</p>
+      <div className="mt-4">
+        <button
+          className="inline-flex justify-center py-2 px-4 border border-transparent text-sm font-medium rounded-md text-white bg-gray-900 hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-gray-500"
+          onClick={resetErrorBoundary}
+        >
+          Reset Query
+        </button>
+      </div>
+    </ErrorState>
+  );
+}
+
+function DBExplorerInner(props: DBExplorerInnerProps) {
+  const { content, extension, filename, sha } = props;
+  const filenameWithoutExtension = filename.split(".").slice(0, -1).join(".");
+  const connectionRef = useRef<duckdb.AsyncDuckDBConnection | null>(null);
+  const [query, setQuery] = useState("");
+  const [debouncedQuery] = useDebounce(query, 500);
+  const [dbStatus, setDbStatus] = useState<"error" | "idle" | "success">(
+    "idle"
+  );
+
+  const execQuery = async (query: string) => {
+    if (!connectionRef.current) return;
+    const queryRes = await connectionRef.current.query(query);
+    const asArray = queryRes.toArray();
+
+    return {
+      numRows: queryRes.numRows,
+      numCols: queryRes.numCols,
+      results: asArray.map((row) => {
+        return row.toJSON();
```

Comment:
This adds some upfront cost that you can save. Arrow tables already work like arrays of objects in most cases.